### PR TITLE
fix(TCOMP-2112): make AvroRecordBuilderFactoryProvider not dependant to jackson-mapper-asl

### PR DESCRIPTION
**Pull request by undx**
https://jira.talendforge.org/browse/TCOMP-2112

- instead of testing a class present in classloader from jackson-mapper-asl, we test an internal class to beam module to be sure that the module is really in classloader and not eventually provided by a far transitive dependency as occurred in described jira issue.

### CODE REVIEW

File: component-runtime-manager/src/test/java/org/talend/sdk/component/runtime/manager/service/record/RecordBuilderFactoryProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.function.Function;
+
+import org.junit.jupiter.api.Test;
+import org.talend.sdk.component.api.service.record.RecordBuilderFactory;
+import org.talend.sdk.component.runtime.manager.ComponentManager;
+import org.talend.sdk.component.runtime.record.RecordBuilderFactoryImpl;
+
+public class RecordBuilderFactoryProviderTest {
```
Comment (clesaec) [inline component-runtime-manager/src/test/java/org/talend/sdk/component/runtime/manager/service/record/RecordBuilderFactoryProviderTest.java:27]: Should be better to enrich real test for [AvroRecordBuilderFactoryProvider](https://github.com/Talend/component-runtime/blob/master/component-runtime-beam/src/test/java/org/talend/sdk/component/runtime/beam/spi/AvroRecordBuilderFactoryProviderTest.java#L30) ? here, just testing a fake class.

To be complete, code 
```java
ofNullable(Thread.currentThread().getContextClassLoader())
                        .orElseGet(ClassLoader::getSystemClassLoader)
                        .loadClass("org.talend.sdk.component.runtime.beam.TalendIO");
```

can be put in a protected method :
```java
protected isTalendPipelineDesigner() {
    try {
        ofNullable(Thread.currentThread().getContextClassLoader())
                        .orElseGet(ClassLoader::getSystemClassLoader)
                        .loadClass("org.talend.sdk.component.runtime.beam.TalendIO");
        return true;
    } catch (final ClassNotFoundException | NoClassDefFoundError cnfe) {
        log
            .info(component-runtime-beam is not available, skipping AvroRecordBuilderFactory ({})",
              getClass().getName());
            return false;
    }
}
```
That make code clearer and easier to test

File: component-runtime-beam/src/main/java/org/talend/sdk/component/runtime/beam/spi/AvroRecordBuilderFactoryProvider.java [hunk 2]
```diff
@@ -48,12 +48,11 @@ public RecordBuilderFactory apply(final String containerId) {
             try {
                 ofNullable(Thread.currentThread().getContextClassLoader())
                         .orElseGet(ClassLoader::getSystemClassLoader)
-                        .loadClass("org.codehaus.jackson.node.TextNode");
+                        .loadClass("org.talend.sdk.component.runtime.beam.TalendIO");
```
[the remark this answers is not in the document]
Comment (ozhelezniak-talend) [reply component-runtime-beam/src/main/java/org/talend/sdk/component/runtime/beam/spi/AvroRecordBuilderFactoryProvider.java:51]: My comment is more about: "Why do we have jackson-mapper-asl" as a test and not compile dependency if component-runtime-beam uses it.

**Review by clesaec — APPROVED** [review]
Nice, 
See comments : code can be made clearer and unit test more accuracy.

File: component-runtime-beam/src/main/java/org/talend/sdk/component/runtime/beam/spi/AvroRecordBuilderFactoryProvider.java [hunk 3]
```diff
@@ -46,14 +46,19 @@ public RecordBuilderFactory apply(final String containerId) {
             return new AvroRecordBuilderFactory(containerId);
         default:
             try {
-                ofNullable(Thread.currentThread().getContextClassLoader())
-                        .orElseGet(ClassLoader::getSystemClassLoader)
-                        .loadClass("org.codehaus.jackson.node.TextNode");
-                return new AvroRecordBuilderFactory(containerId);
+                final ClassLoader cl = ofNullable(Thread.currentThread().getContextClassLoader())
```
Comment (clesaec) [inline component-runtime-beam/src/main/java/org/talend/sdk/component/runtime/beam/spi/AvroRecordBuilderFactoryProvider.java:49]: While you have to rebase, extract the method that chose implementation :
```java
    @Override
    public RecordBuilderFactory apply(final String containerId) {
        switch (System.getProperty("talend.component.beam.record.factory.impl", "auto")) {
        case "memory":
        case "default":
            return new RecordBuilderFactoryImpl(containerId);
        case "avro":
            return new AvroRecordBuilderFactory(containerId);
        default:
            if (this.isAvroImpl()) {
                 return new AvroRecordBuilderFactory(containerId);
            }
            else {
                return new RecordBuilderFactoryImpl(containerId);
            }
        }
    }

protected boolean isAvroImpl() {
    try {
            final ClassLoader cl = ofNullable(Thread.currentThread().getContextClassLoader())
                        .orElseGet(ClassLoader::getSystemClassLoader);
            final Class<?> c1 = cl.loadClass("org.codehaus.jackson.node.TextNode");
            final Class<?> c2 = cl.loadClass("org.talend.sdk.component.runtime.beam.spi.record.AvroSchema");
            return c1.getClassLoader().equals(c2.getClassLoader());
     } catch (final ClassNotFoundException | NoClassDefFoundError cnfe) {
            log.info("component-runtime-beam is not available, skipping AvroRecordBuilderFactory ({}).",
                        getClass().getName());
                return false;
     }
}
```

**Review by clesaec — CHANGES_REQUESTED** [review]
Change asked in same time of rebase

File: component-runtime-manager/src/test/java/org/talend/sdk/component/runtime/manager/service/record/RecordBuilderFactoryProviderTest.java (hunk 1 above)
Comment (undx) [reply component-runtime-manager/src/test/java/org/talend/sdk/component/runtime/manager/service/record/RecordBuilderFactoryProviderTest.java:27]: outdated

File: component-runtime-beam/src/main/java/org/talend/sdk/component/runtime/beam/spi/AvroRecordBuilderFactoryProvider.java (hunk 3 above)
Comment (undx) [reply component-runtime-beam/src/main/java/org/talend/sdk/component/runtime/beam/spi/AvroRecordBuilderFactoryProvider.java:49]: Done +1

**Review by clesaec — APPROVED** [review]
:+1:

**Pull request merged**